src/rtc_pl031.rs

```diff
@@ -0,0 +1,617 @@
+#[derive(Debug)]
+pub enum Error<E> {
+    /// Failed to trigger interrupt.
+    Trigger(E),
+    /// Attempting to write to a non-existent register offset
+    /// or to the match register, which is not emulated.
+    BadWriteOffset(u64),
+}
+
+impl<T: Trigger> RTC<T> {
+    /// Creates a new `AMBA PL031 RTC` instance which uses `trigger` object
+    /// to notify the driver about new events.
+    ///
+    /// # Arguments
+    /// * `trigger` - The Trigger object that will be used to notify the driver
+    ///               about events.
+    ///
+    /// # Example
+    ///
+    /// You can see an example of how to use this function in the
+    /// [`Example` section from `RTC`](struct.RTC.html#example).
+    pub fn new(trigger: T) -> RTC<T> {
+        RTC {
+            // Counts up from 1 on reset at 1Hz (emulated).
+            counter: Instant::now(),
+
+            // The load register is initialized to zero.
+            lr: 0,
+
+            // The match register is initialised to zero (not currently used).
+            // TODO: Implement the match register functionality.
+            mr: 0,
+
+            // The interrupt mask is initialised as not set.
+            imsc: 0,
+
+            // The raw interrupt is initialised as not asserted.
+            ris: 0,
+
+            interrupt_evt: trigger,
+        }
+    }
+
+    /// Provides a reference to the interrupt event object.
+    pub fn interrupt_evt(&self) -> &T {
+        &self.interrupt_evt
+    }
+
+    // TODO: Reimplement with the match register.
+    // fn trigger_interrupt(&mut self) -> Result<(), T::E> {
+    //     // Triggers an event which is sent to the driver.
+    //     self.interrupt_evt.trigger()
+    // }
+
+    fn get_rtc_value(&self) -> u32 {
+        // Add the counter offset to the seconds elapsed since reset.
+        (self.counter.elapsed().as_secs() as u32) + self.lr
+    }
+
+    /// Handles a write request from the driver at `offset` offset from the
+    /// base register address.
+    ///
+    /// # Arguments
+    /// * `offset` - The offset from the base register specifying
+    ///              the register to be written.
+    /// * `data` - The little endian, 4 byte array to write to the register
+    ///
+    /// # Example
+    ///
+    /// You can see an example of how to use this function in the
+    /// [`Example` section from `RTC`](struct.RTC.html#example).
+    pub fn write(&mut self, offset: u64, data: &[u8; 4]) -> Result<(), Error<T::E>> {
+        let val = u32::from_le_bytes(*data);
+
+        match offset {
+            RTCDR => (), // Ingore writes to the data register.
+            RTCMR => {
+                // Set the match register, though this is not currently used.
+                // TODO: Implement the match register functionality.
+                self.mr = val;
+            }
+            RTCLR => {
+                // Writing to the load register changes the load register.
+                self.lr = val;
+            }
```

Comment:
You also need to reset `self.counter` here. Otherwise, reading back `RTCDR` immediately after writing `RTCLR` will not produce the written value, but one that is offset by the number of seconds since the last RTC reset. See the comment further down for a test case.

In the original Firecracker code, the counter is reset by setting `self.previous_now`.

Reply:
This is not clear to me from the Technical Reference Manual.

From section 2.1 (Real Time Clock overview):
```
The 32-bit counter is incremented on successive rising edges of the input clock CLK1HZ.
Counting in one second intervals is achieved by using a 1Hz clock signal for CLK1HZ.
The counter is free-running and cannot be loaded. On reset, the counter:
• Counts up from one.
...
```

In section 2.2.4 (Update block), it distinguishes between the counter, update value (RTCLR), and RTC value (RTCDR):
```
There is a distinction between the RTC value and counter value:
• An update value (through the RTCLR register) is applied to the counter value in the update block.
The resulting offset is applied to the current counter value to generate an updated RTC value.

• Reads from the Data Register, RTCDR, return the current value of the RTC alone and not that of the
counter, as they are different.

Generally, an update to the absolute RTC value occurs after two rising clock edges of PCLK.
Also, an RTC enable bit is set when 1 is written to bit[0] of the Control Register.
When set, the RTC is started but any subsequent write resets the current RTC value.
A read of bit[0] indicates the status of the RTC enable signal.
```
To me, this strongly suggests that the counter is only reset when you write to `RTCCR`.  When you write to `RTCLR`, it updates the offset, which will be applied to the *current* counter value when reading `RTCDR`.

In section 3.3.3 (Load Register, RTCLR), it simply says:
```
RTCLR is a 32-bit read/write load register. Writes to this register load an update value into the RTC Update logic
block where the updated value of the RTC is calculated. Reads return the last written value.
```
There is no mention of counter reset here.

The only mention of resetting the base counter is in section 3.3.4 (Control Register, RTCCR):
```
If set to 1, the RTC is enabled. After the RTC is enabled, do not write to this bit otherwise the
current RTC value is reset to zero.
```

I've tried to implement this as I understand the above:
1. Writes to `RTCCR` reset both the counter and the offset value.
2. Writes to `RTCLR` update the offset, but do not affect the counter value.
3. `RTCCR` counts up from 1 and `RTCLR` is initialised to zero, per the reset values in section 3.2 (Summary of RTC registers).

All three of those diverge from the current Firecracker implementation.

Reply:
> To me, this strongly suggests that the counter is only reset when you write to `RTCCR`. When you write to `RTCLR`, it updates the offset, which will be applied to the _current_ counter value when reading `RTCDR`.

Your interpretation is correct. The invariant is that `read(RTCDR) == counter + offset`, where `counter` keeps steadily ticking upwards at 1 Hz and `offset` remains fixed unless the guest intends to update RTC time.

> I've tried to implement this as I understand the above:
> [...]
>     2. Writes to `RTCLR` update the offset, but do not affect the counter value.

That is the right approach. What's currently missing is that when the guest driver writes a new value to `RTCLR`, we must update the device state so that `read(RTCDR) == RTCLR_new` holds at exactly the time of the write. Whether that happens by resetting `counter` or not is an implementation detail; if it is not reset, then `offset` must be adjusted accordingly.

In your current code, `read(RTCDR) == (self.counter.elapsed().as_secs() as u32) + self.lr`, and writing `RTCLR` updates `self.lr`, but it does not ensure that `read(RTCDR) == RTCLR_new`. I.e. the guest currently does not see the expected value loaded into the RTC. This can be seen in the test case I posted below, where the guest writes `0` to `RTCLR` but reads back `1` from `RTCDR` -- this should not happen, the two should be identical if no time has passed in between the write and the read.

So my suggestion is the following; I could have been more clear in my wording in that we are not exactly resetting `self.counter` to zero, but rather zeroing `self.counter.elapsed()` by assigning the current system time to the counter.

```suggestion
            RTCLR => {
                // Writing to the load register changes the load register.
                self.counter = Instant::now();
                self.lr = val;
            }
```

Reply:
Ah, I'm tracking with you now.  Sorry.  I think you were sufficiently clear from the start and this was a concept error on my part.  I'll make the adjustment.

Reply:
I've just pushed a correction for this.